File: Kakao/2020_Internship/2번.cpp
#include<iostream>
#include <string>
#include <vector>
using namespace std;
char op[] = { '+','-','*' };
bool isCheck[3];
vector<long long> numbers;
vector<char> ops;
long long answer = 0;
long long calc(long long num1, long long num2, char op) {
	long long num = 0;
	switch (op) {
	case '+': {
		return num1 + num2;
	}
	case '-': {
		return num1 - num2;
	}
	case '*': {
		return num1 * num2;
	}
	}
	return num;
}

void dfs(int count, char p[3]) {
	if (count == 3) {
		vector<long long> tempNumbers = numbers;
		vector<char> tempOps = ops;
		for (int i = 0; i < 3; ++i) {
			for (int j = 0; j < tempOps.size(); ++j) {
				if (p[i] == tempOps[j]) {
					long long num1 = tempNumbers[j];
					tempNumbers.erase(tempNumbers.begin() + j);
					long long num2 = tempNumbers[j];
					tempNumbers.erase(tempNumbers.begin() + j);
					long long res = calc(num1, num2, p[i]);
					tempNumbers.insert(tempNumbers.begin() + j, res);
					tempOps.erase(tempOps.begin() + j);
					--j;
				}
			}
		}
		answer = answer > abs(tempNumbers[0]) ? answer : abs(tempNumbers[0]);
		return;
	}
	for (int i = 0; i < 3; i++) {
		if (!isCheck[i]) {
			isCheck[i] = true;
			p[count] = op[i];
			dfs(count + 1, p);
			isCheck[i] = false;
		}
	}
}
long long solution(string expression) {
	answer = 0;
	string num = "";
	for (int i = 0; i < expression.length(); ++i) {
		if (expression[i] >= '0' && expression[i] <= '9') {
			num += expression[i];
		}
		else {
			numbers.push_back(stoll(num));
			num = "";
			ops.push_back(expression[i]);
		}
	}
	numbers.push_back(stoll(num));
	dfs(0, new char[3]);
	return answer;
}

int main()
{
	cout << solution("100-200*300-500+20") << endl;
	cout << solution("50*6-3*2") << endl;
}
/*
"100-200*300-500+20"	60420
"50*6-3*2"	300
*/

File: Kakao/2020_Internship/2번.java
import java.util.*;

public class Solution {
	char[] ops;
	long answer;
	String expression;

	public Long solution(String expression) {
		this.expression = expression;
		List<Character> op = new ArrayList<>();
		if (expression.contains("+")) {
			op.add('+');
		}
		if (expression.contains("-")) {
			op.add('-');
		}
		if (expression.contains("*")) {
			op.add('*');
		}
		ops = new char[op.size()];
		for (int i = 0; i < ops.length; ++i) {
			ops[i] = op.get(i);
		}
		answer = -1;
		DFS(0);
		return answer;
	}

	void DFS(int depth) {
		if (depth == ops.length) {
			long result = calc();
			answer = Math.max(result, answer);
			return;
		}
		for (int i = depth; i < ops.length; ++i) {
			swap(i, depth);
			DFS(depth + 1);
			swap(i, depth);
		}
	}

	void swap(int i, int j) {
		char c = ops[i];
		ops[i] = ops[j];
		ops[j] = c;
	}

	long calc() {
		HashMap<String, Integer> hm = new HashMap<>();
		for (int i = 0; i < ops.length; i++) {
			hm.put(String.valueOf(ops[i]), i);
		}
		StringBuilder sb = new StringBuilder();
		List<String> list = new ArrayList<>();
		Stack<String> arg = new Stack<>();
		int len = this.expression.length();
		for (int i = 0; i < len; i++) {
			char c = this.expression.charAt(i);
			if (c == '+' || c == '-' || c == '*') {
				list.add(sb.toString());
				sb.delete(0, sb.length());
				while (!arg.isEmpty() && hm.get(String.valueOf(c)) <= hm.get(arg.peek())) {
					list.add(arg.pop());
				}
				arg.push(String.valueOf(c));
			} else {
				sb.append(c);
			}
		}
		list.add(sb.toString());
		while (!arg.isEmpty()) {
			list.add(arg.pop());
		}

		Stack<Long> value = new Stack<>();
		for (String s : list) {
			if (hm.containsKey(s)) {
				long b = value.pop();
				long a = value.pop();
				if (s.equals("+")) {
					value.push(a + b);
				}
				if (s.equals("-")) {
					value.push(a - b);
				}
				if (s.equals("*")) {
					value.push(a * b);
				}
			} else {
				value.push(Long.parseLong(s));
			}
		}
		return Math.abs(value.pop());
	}
}
